File: experiments/repair-loop/testkits/002-bank-ledger/python/solution.py
"""Phase-1.5 stub bank-ledger solution for Python adapter validation.

Minimal working implementation: satisfies the 002-bank-ledger API surface
with correct types and basic invariants, passes pyright (strict-mode off)
and the colocated smoke test. Not the canonical Phase-2/3 solution — agents
will write their own. Used here only to validate that the Python target
adapter wires pyright and pytest through the orchestrator without
schema-coupling defects.
"""
from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import Mapping


@dataclass(frozen=True)
class Transfer:
    from_account: str
    to_account: str
    amount: int


@dataclass(frozen=True)
class Ledger:
    balances: Mapping[str, int]
    log: tuple[Transfer, ...] = field(default_factory=tuple)


class LedgerError(Exception):
    """Base class for transfer/balance failures."""


class MissingAccountError(LedgerError):
    pass


class InsufficientFundsError(LedgerError):
    pass


class NonPositiveAmountError(LedgerError):
    pass


def create_ledger(initial: Mapping[str, int]) -> Ledger:
    return Ledger(balances=dict(initial), log=())


def balance(ledger: Ledger, account_id: str) -> int:
    if account_id not in ledger.balances:
        raise MissingAccountError(account_id)
    return ledger.balances[account_id]


def transfer(
    ledger: Ledger, from_account: str, to_account: str, amount: int
) -> Ledger:
    if amount <= 0:
        raise NonPositiveAmountError(amount)
    if from_account not in ledger.balances:
        raise MissingAccountError(from_account)
    if to_account not in ledger.balances:
        raise MissingAccountError(to_account)
    if ledger.balances[from_account] < amount:
        raise InsufficientFundsError(amount)
    new_balances = dict(ledger.balances)
    new_balances[from_account] -= amount
    new_balances[to_account] += amount
    new_log = ledger.log + (Transfer(from_account, to_account, amount),)
    return replace(ledger, balances=new_balances, log=new_log)


def total_balance(ledger: Ledger) -> int:
    return sum(ledger.balances.values())

File: experiments/repair-loop/testkits/002-bank-ledger/go/solution.go
// Phase-1.75 stub bank-ledger solution for Go adapter validation.
//
// Module-mode upgrade from Phase-1.5: paired with go.mod and
// solution_test.go (both harness-owned via targets/go.json
// `harness_files`). Verifier chain runs `go vet ./...`, `go build ./...`,
// `go test ./...`. Not the canonical Phase-2/3 solution — agents will
// write their own.
package main

import (
	"errors"
	"fmt"
)

type Ledger struct {
	balances map[string]int64
	log      []TransferRecord
}

type TransferRecord struct {
	From   string
	To     string
	Amount int64
}

var (
	ErrMissingAccount    = errors.New("account does not exist")
	ErrInsufficientFunds = errors.New("source account has insufficient funds")
	ErrNonPositiveAmount = errors.New("transfer amount must be positive")
)

func CreateLedger(initial map[string]int64) *Ledger {
	bal := make(map[string]int64, len(initial))
	for k, v := range initial {
		bal[k] = v
	}
	return &Ledger{balances: bal, log: nil}
}

func Balance(l *Ledger, accountID string) (int64, error) {
	b, ok := l.balances[accountID]
	if !ok {
		return 0, ErrMissingAccount
	}
	return b, nil
}

func Transfer(l *Ledger, from, to string, amount int64) (*Ledger, error) {
	if amount <= 0 {
		return nil, ErrNonPositiveAmount
	}
	fromBal, ok := l.balances[from]
	if !ok {
		return nil, ErrMissingAccount
	}
	if _, ok := l.balances[to]; !ok {
		return nil, ErrMissingAccount
	}
	if fromBal < amount {
		return nil, ErrInsufficientFunds
	}
	newBalances := make(map[string]int64, len(l.balances))
	for k, v := range l.balances {
		newBalances[k] = v
	}
	newBalances[from] -= amount
	newBalances[to] += amount
	newLog := append([]TransferRecord(nil), l.log...)
	newLog = append(newLog, TransferRecord{From: from, To: to, Amount: amount})
	return &Ledger{balances: newBalances, log: newLog}, nil
}

func TotalBalance(l *Ledger) int64 {
	var sum int64
	for _, v := range l.balances {
		sum += v
	}
	return sum
}

func main() {
	// Smoke main: not exercised by the harness. Exists only so `go build`
	// in package-main mode produces an executable.
	l := CreateLedger(map[string]int64{"alice": 1000, "bob": 500})
	fmt.Println("total:", TotalBalance(l))
}
